include/unifex/exception.hpp

```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright 2019-present Facebook, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <exception>
+#include <utility>
+#include <type_traits>
+
+#include <unifex/std_concepts.hpp>
+
+#include <unifex/detail/prologue.hpp>
+
+// Microsoft's STL has a fast implementation of std::make_exception_ptr.
+// So does libc++ when targetting Microsoft's ABI. On other stdlibs,
+// move as much exception machinery into a cpp file as possible to avoid
+// template code bloat.
+#if !defined(_LIBCPP_ABI_MICROSOFT) && !defined(_MSVC_STL_VERSION)
+#define UNIFEX_HAS_FAST_MAKE_EXCEPTION_PTR() 0
+#else
+#define UNIFEX_HAS_FAST_MAKE_EXCEPTION_PTR() 1
+#endif
+
+namespace unifex {
+namespace _except_ptr {
+
+#if !UNIFEX_HAS_FAST_MAKE_EXCEPTION_PTR()
+// If std::make_exeption_ptr() is slow, then move it into a cpp
+// file to generate less code elsewhere.
+struct _ref {
+  template (typename Obj)
+    (requires (!same_as<remove_cvref_t<Obj>, _ref>))
+  _ref(Obj&& obj) noexcept
+    : p_((void*) std::addressof(obj))
+    , throw_([](void* p) {
+        throw (Obj&&) *(std::add_pointer_t<Obj>) p;
+      })
```

Comment:
I wonder, would using a static function template rather than a lambda here give the linker a better chance at deduplicating the symbols compared to a lambda.

Also, is there any benefit in declaring these as `[[noreturn]]`?
I don't think so, because it's type-erased through a function-pointer and function-pointers can't be declared as `[[noreturn]]`, but maybe I'm missing something?

Reply:
I could replace the lambda with a static function template. Are you asking whether `rethrow()` should be declared `[[noreturn]]`?

Reply:
I was asking whether the lambda body should be declared as `[[nothrow]]`- I'm not sure if this is currently possible...
But making it a static function might make it possible.

e.g
```
private:
  template<typename Obj>
  [[noreturn]] void throw_impl(void* p) {
    throw static_cast<Obj&&>(*static_cast<std::add_pointer_t<Obj>>(p));
  }
public:
  template<typename Obj>
  ref(Obj&& obj) noexcept
  : p_(static_cast<void*>(std::addressof(obj)))
  , throw_(&throw_impl<Obj>) {}

  ...
};
```

But it seems you can't declare a function signature type as `[[noreturn]]` so the callsite is not going to be able to deduce this anyway (which you've handled with the UNIFEX_ASSUME_UNREACHABLE() macro).

Reply:
Yes, attributes aren't part of the type system, so putting one here would have no effect.